src/doorman/Channel.cpp

```diff
@@ -0,0 +1,393 @@
+    size_t i;
+    for (i = 0; i < (sizeof(terms) / sizeof(terms[0])); ++i)
+        if (!strcasecmp(src, terms[i]))
+            return true;
+    return false;
+}
+
+void Channel::closeConnection() {
+    // Log the source IP but masked for privacy.
+    log_out("[%d] Closing connection to %s", id_, get_masked_hostname(hostname_).c_str());
+    close();
+}
+
+void Channel::set_echo(bool echo) {
+    SendCom(fd_, echo ? WONT : WILL, TELOPT_ECHO);
+    echoing_ = echo;
+}
+
+void Channel::on_reconnect_attempt() {
+    if (!fd_.is_open())
+        return;
+    if (!sent_reconnect_message_) {
+        send_to_client("Attempting to reconnect to Xania"sv);
+        sent_reconnect_message_ = true;
+    }
+    send_to_client(".", 1);
+}
+
+void Channel::close() {
+    fd_.close();
+    host_lookup_fd_.close();
+    if (host_lookup_pid_) {
+        if (::kill(host_lookup_pid_, 9) < 0)
+            log_out("Couldn't kill child process (ident process has already exited)");
+        host_lookup_pid_ = 0;
+    }
+}
+
+void Channel::newConnection(Fd fd, sockaddr_in address) {
+    // One day I will be a constructor...
+    this->fd_ = std::move(fd);
+    hostname_ = inet_ntoa(address.sin_addr);
+    port_ = ntohs(address.sin_port);
+    netaddr_ = ntohl(address.sin_addr.s_addr);
+    buffer_.clear();
+    width_ = 80;
+    height_ = 24;
+    ansi_ = got_term_ = false;
+    echoing_ = true;
+    sent_reconnect_message_ = false;
+    connected_ = false;
+    auth_char_name_.clear();
+
+    log_out("[%d] Incoming connection from %s on fd %d", id_, get_masked_hostname(hostname_).c_str(), fd_.number());
+
+    // Send all options out
+    sendTelopts();
+
+    // Tell them if the mud is down
+    if (!mud_->connected()) {
+        auto msg = "Xania is down at the moment - you will be connected as soon "
+                   "as it is up again.\n\r"sv;
+        send_to_client(msg.data(), msg.size());
+    }
+
+    sendConnectPacket();
+
+    // Start the async hostname lookup process.
+    int pipe_fds[2];
+    if (pipe(pipe_fds) == -1) {
+        log_out("Pipe failed!");
+        return;
+    }
+    Fd parent_fd(pipe_fds[0]);
+    Fd child_fd(pipe_fds[1]);
+    log_out("Pipes %d<->%d", pipe_fds[0], pipe_fds[1]);
+    auto forkRet = fork();
+    switch (forkRet) {
+    case 0: {
+        // Child process - go and do the lookup
+        log_out("ID: doing lookup");
+        async_lookup_process(address, std::move(child_fd));
+        log_out("ID: exiting");
+        exit(0);
+    } break;
+    case -1:
+        // Error - unable to fork()
+        log_out("Unable to fork() an IdentD lookup process");
+        break;
+    default:
+        host_lookup_fd_ = std::move(parent_fd);
+        host_lookup_pid_ = forkRet;
+        log_out("Forked hostname process has PID %d on %d", forkRet, host_lookup_fd_.number());
+        break;
+    }
+}
+
+bool Channel::incomingData(gsl::span<const byte> incoming_data) {
```

Comment:
aha,  the modern way to do char* foo, size_t n  :)

Reply:
> aha, the modern way to do char* foo, size_t n :)

In this case `const byte *, size` yeah. 

Aside: There's a language-level thing for strings: `string_view` is `const char*, size_t` and I use it in a bunch of places. It can be automatically converted-to from a `const char *` and a `std::string` so it's a good way to pass a non-owning reference to a string. BUT it's not NUL terminated, so only useful if you don't need that.